[CK_TILE] Epilogue chaining (Lwpck 3373)

amd-yashagar opened the pull request:
## Proposed changes

The EpilogueChainer allows to stack epilogues one after the other and run them sequentially.

The EpilogueChainer implementation is demonstrated using the cshuffle epilogue by breaking and chaining it. The breaking and testing of the cshuffle epilogue is taken from #2729, hence this also has the row/col scaling. The example shows how the epilogues are modular and can sort of plugged-in if needed.

## Checklist

Please put an `x` into the boxes that apply. You can also fill these out after creating the PR. If you're not sure, please don't hesitate to ask.

- [ ] I have added tests relevant to the introduced functionality, and the unit tests are passing locally
- [ ] I have added the test to REGRESSION_TESTS list defined at the top of CMakeLists.txt in tests/CMakeLists.txt, **IF** the test takes more than 30 seconds to run.
- [ ] I have added inline documentation which enables the maintainers with understanding the motivation
- [ ] I have removed the stale documentation which is no longer relevant after this pull request
- [ ] (If this change is user-facing) I have added release notes which provide the end users with a brief summary of the improvement from this pull request
- [ ] I have run `clang-format` on all changed files
- [ ] Any dependent changes have been merged

## Discussion

If this is a relatively large or complex change, feel free to start a discussion by explaining why you chose the solution you did and what alternatives you considered

amd-yashagar committed:
Epilogue chainer

* epilogue chainer with context to share state in between epilogues
* chain-able epilogues for cshuffle

amd-yashagar committed:
clang-format

amd-yashagar committed:
rebase related changes

- Added separate chainer test
-  clang format

mozga-amd commented on include/ck_tile/ops/epilogue/cshuffle_chained_epilogues.hpp:
```diff
@@ -0,0 +1,573 @@
+// SPDX-License-Identifier: MIT
+// Copyright (c) 2018-2025, Advanced Micro Devices, Inc. All rights reserved.
```
2025

mozga-amd commented on include/ck_tile/ops/epilogue/chainer/cshuffle_epilogue_chainer_ops.hpp:
```diff
@@ -0,0 +1,573 @@
+    };
+
+    template <typename ODramWindow, typename OAccTile, typename DsDramWindows>
+    CK_TILE_DEVICE auto operator()([[maybe_unused]] ODramWindow& out_dram_window,
```
`[[maybe_unused]]` Is it necessary?

mozga-amd commented on include/ck_tile/ops/epilogue/cshuffle_chained_epilogues.hpp:
```diff
@@ -0,0 +1,573 @@
+                                   [[maybe_unused]] auto iAccess,
+                                   [[maybe_unused]] ContextType& context)
+    {
+        (void)out_dram_window;
```
leftovers?

mozga-amd commented on include/ck_tile/ops/epilogue/chainer/cshuffle_chained_epilogues.hpp:
```diff
@@ -0,0 +1,573 @@
+    static constexpr index_t NPerIterationShuffle     = Base::NPerIterationShuffle;
+
+    template <typename ODramWindow, typename OAccTile, typename DsDramWindows, typename ContextType>
+    CK_TILE_DEVICE auto operator()([[maybe_unused]] ODramWindow& out_dram_window,
```
[[maybe_unused]] - It's necessary here and below?

mozga-amd commented on include/ck_tile/ops/epilogue/chainer/cshuffle_chained_epilogues.hpp:
```diff
@@ -0,0 +1,573 @@
+              typename ContextType,
+              typename ScaleM,
+              typename ScaleN>
+    CK_TILE_DEVICE auto operator()([[maybe_unused]] ODramWindow& out_dram_window,
```
[[maybe_unused]] - It's necessary here and below?

mozga-amd commented on include/ck_tile/ops/epilogue/chainer/cshuffle_chained_epilogues.hpp:
```diff
@@ -0,0 +1,573 @@
+        const auto scale_m_tile = load_tile(scale_m_window);
+        const auto scale_n_tile = load_tile(scale_n_window);
+
+        tile_elementwise_inout(element_wise::MultiDMultiply{},
```
Can we only use the elementwise multiplication operator? What about other elementwise operators — can they occur as well?

amd-yashagar replied:
yeah other operators should also be usable. This specific epilogue is for the quant scaling, but new epilogue can be created or this one can be modified into a general one with option  to specify the operator for elementwise.

mozga-amd replied:
Would it be better if this elementwise function were passed in and configured externally, and this one has a more general?

mozga-amd commented on include/ck_tile/ops/epilogue/chainer/cshuffle_chained_epilogues.hpp:
```diff
@@ -0,0 +1,573 @@
+    using ODataType = typename Base::ODataType;
+
+    template <typename ODramWindow, typename OAccTile, typename DsDramWindows, typename ContextType>
+    CK_TILE_DEVICE auto operator()([[maybe_unused]] ODramWindow& out_dram_window,
```
[[maybe_unused]] - It's necessary here and below?

mozga-amd commented on include/ck_tile/ops/epilogue/chainer/cshuffle_chained_epilogues.hpp:
```diff
@@ -0,0 +1,573 @@
+    static constexpr index_t NumDTensor = Base::NumDTensor;
+
+    template <typename ODramWindow, typename OAccTile, typename DsDramWindows, typename ContextType>
+    CK_TILE_DEVICE auto operator()([[maybe_unused]] ODramWindow& out_dram_window,
```
[[maybe_unused]] - It's necessary here and below?

mozga-amd commented on include/ck_tile/ops/epilogue/cshuffle_chained_epilogues.hpp:
```diff
@@ -0,0 +1,573 @@
+    static constexpr memory_operation_enum MemoryOperation = Base::MemoryOperation;
+
+    template <typename ODramWindow, typename OAccTile, typename DsDramWindows, typename ContextType>
+    CK_TILE_DEVICE auto operator()([[maybe_unused]] ODramWindow& out_dram_window,
```
[[maybe_unused]] - It's necessary here and below?

mozga-amd commented on include/ck_tile/ops/epilogue/cshuffle_chained_epilogues.hpp:
```diff
@@ -0,0 +1,573 @@
+    static constexpr index_t NumDTensor = Base::NumDTensor;
+
+    template <typename ODramWindow, typename OAccTile, typename DsDramWindows, typename ContextType>
+    CK_TILE_DEVICE auto operator()([[maybe_unused]] ODramWindow& out_dram_window,
```
[[maybe_unused]] - It's necessary here and below?

mozga-amd commented on include/ck_tile/ops/epilogue/epilogue_chainer.hpp:
```diff
@@ -0,0 +1,284 @@
+// SPDX-License-Identifier: MIT
+// Copyright (c) 2018-2025, Advanced Micro Devices, Inc. All rights reserved.
```
2025

mozga-amd commented on include/ck_tile/ops/epilogue/cshuffle_chained_epilogues.hpp:
```diff
@@ -0,0 +1,573 @@
+                static_assert((kMPerBlock % (MPerXdl * MWave) == 0) &&
+                                  (kMPerBlock % num_xdl_shuffles == 0),
+                              "kMPerBlock must be divisible by MPerXdl*MWave and "
+                              "num_xdl_shuffles for CShuffleEpilogue");
```
CShuffleEpilogue - > CShuffleEpilogueStageBase ?

mozga-amd commented on include/ck_tile/ops/epilogue/cshuffle_chained_epilogues.hpp:
```diff
@@ -0,0 +1,573 @@
+                static_assert((kNPerBlock % (NPerXdl * NWave) == 0) &&
+                                  (kNPerBlock % num_xdl_shuffles == 0),
+                              "kNPerBlock must be divisible by NPerXdl*NWave and "
+                              "num_xdl_shuffles for CShuffleEpilogue");
```
CShuffleEpilogue - > CShuffleEpilogueStageBase ?

mozga-amd commented on include/ck_tile/ops/epilogue/chainer/cshuffle_chained_epilogues.hpp:
```diff
@@ -0,0 +1,573 @@
+};
+
+template <typename Problem_>
+struct ScaleEpilogue : public CShuffleEpilogueStageBase<Problem_>
```
Do we need to use inheritance if we’re already injecting the base class into `using Base = CShuffleEpilogueStageBase<Problem_>;`?

amd-yashagar replied:
We need the inheritance because some of the methods in the base class maybe accessed through the derived class.

mozga-amd commented on include/ck_tile/ops/epilogue/cshuffle_chained_epilogues.hpp:
```diff
@@ -0,0 +1,573 @@
+                                   [[maybe_unused]] void* p_smem)
+    {
+        static_assert(std::is_same_v<ELayout, tensor_layout::gemm::RowMajor>,
+                      "Currently, the CShuffle Epilogue only supports the Row Major Output layout");
```
```suggestion
                      "Currently, the CShuffleEpilogueStageBase only supports the Row Major Output layout");
```

mozga-amd reviewed (commented):
Looks good overall! Nice work.

amd-yashagar committed:
comment resolutions

amd-yashagar committed:
clang-format

amd-yashagar committed:
Merge branch 'develop' into LWPCK-3373

AviralGoelAMD reviewed (dismissed):
Good addition! 
Can we document this change as a write up in README.md in the test/ck_tile/epilogue to explain novice users how the chaining can be done? (the @brief in the comments is not enough)

AviralGoelAMD commented:
PR also seems a good candidate to be added to CHANGELOG.md

amd-yashagar committed:
Merge branch 'develop' into LWPCK-3373

amd-yashagar committed:
Policy based chaining

- basic Policy structure to control blanket looping and barrier
placement.

- to be extended for fine grianed control

- to  be modified to move possible auto-compute values and SFC  access
count to policy

amd-yashagar committed:
Refactoring as per spec

- Introduced epilogue schedule, graph
- modified chainer to function with graph and schedule

amd-yashagar committed:
minor_changes

- made functions to overload in the epilogue_graph file

amd-yashagar committed:
clang-format

amd-yashagar committed:
Merge branch 'develop' into LWPCK-3373

amd-yashagar committed:
Documentation and Comments

- Added comments to files
- Noted changes in changelog
- Added README to explain the chainer and current status, exact use
steps to be added

spolifroni-amd commented on include/ck_tile/ops/epilogue/chainer/README.md:
```diff
@@ -0,0 +1,48 @@
+
+## Overview
+
+The Epilogue Chainer provides composability for building epilogue sequences for GEMM operations. It allows to chain multiple epilogue stages to be executed sequentially.
```
```suggestion
The Epilogue Chainer provides composability for building epilogue sequences for GEMM operations. It provides a means to chain multiple epilogue stages so they can be run sequentially.
```

I switched "executed" to "run" because "execute" is one of the "words to avoid".

spolifroni-amd commented on include/ck_tile/ops/epilogue/chainer/README.md:
```diff
@@ -0,0 +1,48 @@
+## Components
+
+### EpilogueChainer
+EpilogueChainer class, orchestrator that executes epilogue sequences. Takes an initialization epilogue and a sequence of operations to perform.
```
```suggestion
The `EpilogueChainer` class is an orchestrator that runs epilogue sequences. It takes an initialization epilogue and a sequence of operations.
```

spolifroni-amd commented on include/ck_tile/ops/epilogue/chainer/README.md:
```diff
@@ -0,0 +1,48 @@
+EpilogueChainer class, orchestrator that executes epilogue sequences. Takes an initialization epilogue and a sequence of operations to perform.
+
+### EpilogueNode
+Wrapper for individual epilogue operations with embedded arguments. Supports both parameterized and parameter-free epilogues.
```
```suggestion
`EpilogueNode` is a wrapper for individual epilogue operations with embedded arguments. Supports both parameterized and parameter-free epilogues.
```

spolifroni-amd commented on include/ck_tile/ops/epilogue/chainer/README.md:
```diff
@@ -0,0 +1,48 @@
+Wrapper for individual epilogue operations with embedded arguments. Supports both parameterized and parameter-free epilogues.
+
+### EpilogueLoop
+Executes a sequence of epilogue nodes across multiple iterations (e.g., for tiled processing).
```
```suggestion
`EpilogueLoop runs a sequence of epilogue nodes across multiple iterations (e.g., for tiled processing).
```

Is this a function? So far we have a class and a wrapper, and it makes sense to say what this is.

spolifroni-amd reviewed (changes requested):
The readme needs some small changes.

amd-yashagar committed:
Comment resolutions

- README modified with the suggested changes
- Comment fixed accordingly

amd-yashagar commented:
@spolifroni-amd Thank you for the review, I have made the suggested changes.

aosewski commented on include/ck_tile/ops/epilogue/chainer/epilogue_graph.hpp:
```diff
@@ -0,0 +1,144 @@
+    {
+        ck_tile::apply(
+            [&](auto&&... epilogue_args) {
+                EpilogueType{}(out_dram_window,
```
You have an alias for `EpilogueType`

aosewski commented on include/ck_tile/ops/epilogue/chainer/epilogue_graph.hpp:
```diff
@@ -0,0 +1,144 @@
+    constexpr EpilogueLoop(EpilogueTypes... eps) : epilogues(eps...) {}
+
+    /// @brief Execute all epilogues for each iteration in sequence
+    template <typename ODramWindow, typename OAccTile, typename DsDramWindows, typename Context>
```
Please explain what is Context type and how it can be used ?

aosewski commented on include/ck_tile/ops/epilogue/chainer/epilogue_chainer.hpp:
```diff
@@ -0,0 +1,64 @@
+class EpilogueChainer
+{
+    public:
+    using SelectEpilogue                  = InitEpilogue;
```
What is this alias meant for ?

aosewski commented on include/ck_tile/ops/epilogue/chainer/epilogue_chainer.hpp:
```diff
@@ -0,0 +1,64 @@
+                                   const OAccTile& o_acc_tile,
+                                   const DsDramWindows& ds_dram_windows,
+                                   void* p_smem,
+                                   const MainSequence& sequence) const
```
Seems to me like this `MainSequence` could be a parameter of just this function call operator.

aosewski commented on include/ck_tile/ops/epilogue/chainer/epilogue_chainer.hpp:
```diff
@@ -0,0 +1,64 @@
+
+namespace ck_tile {
+
+/// @brief Orchestrates execution of epilogue sequences
```
I wondering whether this is a good explanation of this class functionality/responsibility. The action of "orchestration" is closer to scheduling for me - sth like it takes multiple different objects and knows in what order and how to run them. However this class is rather providing context for all executed epilogues stages and just run them with it. 
Additionally I'm wondering whether we shouldn't provide any safeguard checks to make sure that we can run whole sequence with provided context in a given order.

aosewski commented on include/ck_tile/ops/epilogue/chainer/cshuffle_epilogue_chainer_ops.hpp:
```diff
@@ -0,0 +1,572 @@
+        {
+            static_assert(false, "Unsupported DLayout!");
+        }
+        return max_vector_size / sizeof(DiDataType);
```
Unreachable code?

aosewski commented on test/ck_tile/epilogue/test_cshuffle_epilogue.cpp:
Why you need to add this here?

aosewski commented on test/ck_tile/epilogue/test_epilogue_chainer_util.hpp:
```diff
@@ -0,0 +1,204 @@
+        test_epilogue_chainer_kernel<Problem, M, N, true>
+            <<<gridSize, blockSize>>>(device_output, m_scale, n_scale);
+
+        HIP_CHECK_ERROR(hipFree(m_scale));
```
If you free device memory before the end of a kernel you may get segfaults.

aosewski commented on test/ck_tile/epilogue/test_epilogue_chainer_util.hpp:
```diff
@@ -0,0 +1,204 @@
+    HIP_CHECK_ERROR(hipMemcpy(
+        host_output.data(), device_output, output_size * sizeof(ODataType), hipMemcpyDeviceToHost));
+
+    // Basic verification
+    bool has_2 =
+        type_convert<float>(host_output[0]) > 1.9F && type_convert<float>(host_output[0]) < 2.1F;
+    bool scale_has_4 = true;
+    if(use_scale)
+    {
+        scale_has_4 = type_convert<float>(host_output[1]) > 3.9F &&
+                      type_convert<float>(host_output[1]) < 4.1F;
+    }
+
```
Please do more thorough results verification.

aosewski commented on test/ck_tile/epilogue/test_epilogue_chainer_util.hpp:
```diff
@@ -0,0 +1,204 @@
+
+    // Fill acc_tile with a simple pattern
+    auto& acc_buffer = acc_tile.get_thread_buffer();
+    acc_buffer[0]    = 2.0F;
```
Please initialize whole tile with some initial value.

aosewski commented on include/ck_tile/ops/epilogue/chainer/epilogue_schedule.hpp:
```diff
@@ -0,0 +1,50 @@
+
+    /// @brief Standard epilogue without scaling
+    /// Sequence: Slice -> Cast -> PrepC -> ApplyD -> Store -> MoveWindows
+    CK_TILE_DEVICE static auto make_base_schedule()
```
IMHO shorter name would be sufficient here. 
```suggestion
    CK_TILE_DEVICE static auto make_schedule()
```

aosewski reviewed (commented):
Great work and great start!

AviralGoelAMD commented on test/ck_tile/epilogue/test_cshuffle_epilogue.cpp:
```diff
@@ -114,6 +115,76 @@ TEST_F(CShuffleEpilogueTest, BasicHalfTestWithTensorScale)
+                                                     NPerXdl,
+                                                     KPerXdl>;
+
+    bool result = run_epilogue_chainer_test<TestProblem, kMPerBlock, kNPerBlock>();
```
to easily compare difference with the next test
```suggestion
    bool result = run_epilogue_chainer_test<TestProblem, kMPerBlock, kNPerBlock>(false);
```

AviralGoelAMD commented on test/ck_tile/epilogue/test_epilogue_chainer_util.hpp:
```diff
@@ -0,0 +1,204 @@
+        auto schedule = Scheduler::make_base_schedule();
+
+        // Execute
+        EpilogueChainer<InitEpilogue, decltype(schedule)>{}(
```
the parameter `schedule` is renamed as `sequence` in the `EpilogueChainer`. This is confusing. Let us stick to one naming convention.

aosewski commented on include/ck_tile/ops/epilogue/chainer/epilogue_schedule.hpp:
One more thought on this file which I forgot to add earlier. This file should be renamed to `cshuffle_epilogue_schedule.hpp` cause it's not a generic epilogue schedule but specific to cshuffle one. We would rather put each specific epilogue schedule into different file.

amd-yashagar committed:
major refactoring

- modified the chainer files to match the new design
- updated comments
- updated readme
- multi-d example shocases use of the chainer

amd-yashagar committed:
Merge branch 'develop' into LWPCK-3373

amd-yashagar committed:
minor cleanup

amd-yashagar committed:
tensor and rowcol quant chainer epilogue

- added scalarepilogue for tensor quant
- added schedule for tensorquant
- modified quant example to use chainer and appropriate schedules

ThomasNing commented on example/ck_tile/19_gemm_multi_d/gemm_multi_d_fp16.cpp:
```diff
@@ -104,24 +106,51 @@ auto gemm_multi_d(const gemm_multi_d_kargs& args, const ck_tile::stream_config&
-                                                 memory_operation>>;
+            // Epilogue selection: set to true for chainer-based, false for standard
+            // CShuffleEpilogue
+            constexpr bool UseChainerEpilogue = false;
```
It's good to know that the Epilogue Chainer is working on the current ck tile example. But let's have another PR and probably enhance our access to the Epilogue Chainer in another PR. S So let's ignore the example files change.

aosewski replied:
@ThomasNing  Could you elaborate what would you like to see? Easier usage or ..  ? This code is not breaking existing examples and duplicating whole example just to introduce chainer isn't a good idea IMHO.

ThomasNing replied:
@aosewski Sorry for the confusion. Here is my intention. Currently, the Epilogue Chainer is bundled in here with the CshuffleEpilogueSchedule and ChainProblem. It is better to rethink on we are going to access the Epilogue chainer in the example or a kernel intialization. A more clear structure would be EpilogueChainer -> EpilogueScheduleofOperatorSchedule in the schedule we will determine on if it is CShuffle or Generic only in the memory output part. They could also share with the same Problem structure.

But we could first have a CShuffle operation chaining usage and the test in this PR and have a better access from the example and kernel level in the next PR.

aosewski replied:
Will address this in AICK-486

illsilin committed:
Merge branch 'develop' into LWPCK-3373

aosewski commented on include/ck_tile/ops/epilogue/chainer/common_epilogue_ops.hpp:
```diff
@@ -0,0 +1,256 @@
+///
+/// @par Overview
+///     This file provides epilogue operations that can be composed
+///     into epilogue graphs.
```
Shouldn't this be above the namespace declaration ?

aosewski commented on include/ck_tile/ops/epilogue/chainer/common_epilogue_ops.hpp:
```diff
@@ -0,0 +1,256 @@
+namespace ck_tile {
+
+/// @file common_epilogue_ops.hpp
+/// @brief Reusable epilogue operations for chainer composition
```
```suggestion
/// @brief Reusable simple epilogue operations which might be used to compose more complex one. 
```

aosewski commented on include/ck_tile/ops/epilogue/chainer/common_epilogue_ops.hpp:
```diff
@@ -0,0 +1,256 @@
+                                   auto iAccess,
+                                   ContextType& context)
+    {
+        block_sync_lds();
```
I think this is unnecessary here. You're not touching LDS data here.

aosewski commented on include/ck_tile/ops/epilogue/chainer/common_epilogue_ops.hpp:
```diff
@@ -0,0 +1,256 @@
+                                   [[maybe_unused]] const DsDramWindows& ds_dram_windows,
+                                   [[maybe_unused]] void* p_smem,
+                                   auto iAccess,
+                                   ContextType& context)
```
What about having a concrete context type, just for this epilogue op? ie. `SliceEpilogueContext`.

aosewski commented on include/ck_tile/ops/epilogue/chainer/common_epilogue_ops.hpp:
```diff
@@ -0,0 +1,256 @@
+        constexpr auto c_warp_y_index_zeros = uniform_sequence_gen_t<CWarpDstr::NDimY, 0>{};
+
+        // Extract the slice from accumulator tile and store in context LDS tile
+        context.lds_tile.get_thread_buffer() = o_acc_tile.get_y_sliced_thread_data(
```
I'd make it more clear, because I was confused by the naming: the `lds_tile` where you store data is in registers.

aosewski commented on include/ck_tile/ops/epilogue/chainer/common_epilogue_ops.hpp:
```diff
@@ -0,0 +1,256 @@
+          index_t NumNXdlPerWavePerShuffle,
+          index_t MPerIterationShuffle,
+          index_t NPerIterationShuffle>
+struct SliceEpilogue
```
Please add documentation to this operation, what it does etc.

aosewski commented on include/ck_tile/ops/epilogue/chainer/common_epilogue_ops.hpp:
```diff
@@ -0,0 +1,256 @@
+
+    template <typename ODramWindow, typename OAccTile, typename DsDramWindows, typename ContextType>
+    CK_TILE_DEVICE auto operator()([[maybe_unused]] ODramWindow& out_dram_window,
+                                   const OAccTile& o_acc_tile,
```
What about renaming the params to be more generic ones? - not specifically tied up to cshuffle epilogue but rather related to the operation itself ? Like this one could be `in_tile`

aosewski commented on include/ck_tile/ops/epilogue/chainer/common_epilogue_ops.hpp:
```diff
@@ -0,0 +1,256 @@
+        const auto c_warptile_in_tensor_casted = cast_tile<ODataType>(context.lds_tile);
+        store_tile(context.in_lds_window, c_warptile_in_tensor_casted);
+    }
+};
```
The naming of this epilogue does not exactly match to what it actually do. It stores the data to LDS as well.
I'd think of rather more generic version of this one, like `CastAndWriteLDS`... or even better would be to have two epilogues out of this one: CastEpilogue & WriteLDS....

aosewski commented on include/ck_tile/ops/epilogue/chainer/common_epilogue_ops.hpp:
I think that we should strive to have in this file really generic ops, out of which one could build customized epilogue. Cause currently those are not generic ones - it would be hard to use them anywhere else except cshuffle epilogue. Additionally the naming should reflect what is exactly done.

aosewski commented on include/ck_tile/ops/epilogue/chainer/cshuffle_epilogue_schedule.hpp:
```diff
@@ -0,0 +1,122 @@
+///     Each tag corresponds to a pre-built schedule, these are used to select a schedule
+struct DefaultScheduleTag
+{
+}; ///< Standard epilogue schedule: Slice → Cast → PrepC → ApplyD → Store → Move
```
Shouldn't this comment appear right above the struct name ?

aosewski commented on include/ck_tile/ops/epilogue/chainer/cshuffle_epilogue_schedule.hpp:
```diff
@@ -0,0 +1,122 @@
+        }
+        else
+        {
+            static_assert(sizeof(ScheduleTag) == 0, "Unknown schedule tag");
```
Why checking size ? IMHO this wouldn't work. Wouldn't this work: 
```suggestion
            static_assert(false, "Unknown schedule tag");
```

aosewski commented on include/ck_tile/ops/epilogue/chainer/epilogue_chainer.hpp:
```diff
@@ -0,0 +1,214 @@
+};
+
+/// @brief Specialization for epilogue operation wrapper with no arguments
+template <typename EpilogueType>
+struct EpilogueNode<EpilogueType>
+{
+    using Epilogue = EpilogueType;
+    ck_tile::tuple<> args;
+
+    constexpr EpilogueNode() = default;
+
+    template <typename ODramWindow, typename OAccTile, typename DsDramWindows, typename Context>
+    CK_TILE_DEVICE void operator()(ODramWindow& out_dram_window,
+                                   const OAccTile& o_acc_tile,
+                                   const DsDramWindows& ds_dram_windows,
+                                   void* p_smem,
+                                   Context& context) const
+    {
+        EpilogueType{}(out_dram_window, o_acc_tile, ds_dram_windows, p_smem, context);
+    }
+
+    template <typename ODramWindow,
+              typename OAccTile,
+              typename DsDramWindows,
+              typename Context,
+              index_t I>
+    CK_TILE_DEVICE void operator()(ODramWindow& out_dram_window,
+                                   const OAccTile& o_acc_tile,
+                                   const DsDramWindows& ds_dram_windows,
+                                   void* p_smem,
+                                   Context& context,
+                                   number<I> iAccess) const
+    {
+        EpilogueType{}(out_dram_window, o_acc_tile, ds_dram_windows, p_smem, iAccess, context);
+    }
+};
```
This isn't a generic epilogue node. It's highly tied up to the cshuffle epilogue functionality. I think we should change it, or at least for now maybe change the naming of the node to `CShuffleEpilogueNode`.

aosewski reviewed (changes requested):
My main concern is that in current state this is all highly tied up to cshuffle operation. The operations (which are less likely to be common but rather chsuffle oriented), the nodes which forces the cshuffle parameters through its signature... I think that for now we might at least change the names so they would reflect that this is cshuffle specific and in next steps we should make this solution more generic.

aosewski committed:
Merge branch 'develop' into LWPCK-3373

aosewski commented:
Just converted to draft as we still discuss the design.

amd-yashagar committed:
Merge branch 'develop' into LWPCK-3373

amd-yashagar committed:
Refactor epilogue chainer: generalize ops and standardize context interface

Address review comments.

Changes:
- Rename CastToLdsOp to CastAndStoreToLdsOp for clarity
- Standardize context member names (working_tile, out_tile, aux_windows)
- Update README documentation with correct operation names
- Clean up parameter naming in epilogue_chainer.hpp (OutWindow, AccTile,
AuxWindows)
- common_epilogue_ops.hpp: General-purpose ops (ScaleScalarOp,
CastAndStoreToLdsOp,
  LoadFromLdsOp, ElementwiseOp, StoreOp, MoveWindowsOp)
- cshuffle_epilogue_chainer_ops.hpp: CShuffle-specific context and slice
operations
- epilogue_chainer.hpp: Cleaned up parameter naming for generality
- Removed test files that are no longer needed. These were added for
intermediate use

amd-yashagar committed:
Merge branch 'develop' into LWPCK-3373

amd-yashagar committed:
update cshuffle chainer ops file w.r.t cshuffle_epilogue.hpp updates & add chainer to quant gemm example

amd-yashagar committed:
Merge branch 'develop' into LWPCK-3373

amd-yashagar committed:
fix compile errors

- CI uses c++17 while the code had c++20 features

amd-yashagar committed:
Merge branch 'develop' into LWPCK-3373

illsilin committed:
Merge branch 'develop' into LWPCK-3373

aosewski commented:
@amd-yashagar  Could you please resolve conflicts?

amd-yashagar committed:
Merge branch 'develop' into LWPCK-3373

amd-yashagar committed:
Merge branch 'develop' into LWPCK-3373
